File: typescript/src/clause.ts
import type { MatchResult } from './matchResult.js';
import type { Parser } from './parser.js';

// -----------------------------------------------------------------------------------------------------------------

/**
 * Base class for all grammar clauses.
 */
export abstract class Clause {
  /**
   * Match this clause at the given position.
   */
  abstract match(parser: Parser, pos: number, bound?: Clause): MatchResult;

  /**
   * Check that all rule references in this clause are valid.
   */
  abstract checkRuleRefs(grammarMap: Map<string, Clause>): void;

  toString(): string {
    return this.constructor.name;
  }
}

File: java/src/main/java/com/squirrelparser/Clause.java
package com.squirrelparser;

import java.util.Map;

/**
 * Base class for all grammar clauses.
 */
public abstract sealed class Clause permits Terminal, HasOneSubClause, HasMultipleSubClauses, Ref {

    /**
     * Match this clause at the given position.
     *
     * @param parser The parser instance
     * @param pos    The position in the input
     * @param bound  Optional bound clause for recovery
     * @return The match result
     */
    public abstract MatchResult match(Parser parser, int pos, Clause bound);

    /**
     * Check that all rule references in this clause are valid.
     *
     * @param grammarMap The map of rule names to clauses
     */
    public abstract void checkRuleRefs(Map<String, Clause> grammarMap);

    @Override
    public String toString() {
        return getClass().getSimpleName();
    }
}
